[v6r15] Conditional FileCatalog instantiation

chaen opened the pull request:
This PR contains the mechanism to allow conditional FC instantiation. It actually does not call it yet. The idea is that a FC is not necessarily to be used in all the cases. This system allows to define conditions in the CS for each FC. The conditions are evaluated, and if true, we use this catalog. The conditions are described using boolean expression, and the basic blocks are made of plugins. This PR comes with two plugins: one to perform test on the LFN, one for performing test on the Proxy. VO can add their own plugins in their extension

andresailer commented on Resources/Catalog/FCConditionParser.py:
```diff
@@ -0,0 +1,312 @@
+      """ Perform the evaluation of the boolean logic
+          by applying the operator between the two arguments
+
+          :param **kwargs whatever information is given to plugin (typically lfn)
```
This :param **kwargs needs to end with a colon.
I.e.:
```
:param **kwargs:
```
There are more cases of missing colons for :param and :return(s).
Otherwise things are not rendered correctly (e.g., http://diracgrid.org/files/docs/CodeDocumentation/Core/Utilities/ReturnValues.html)

chaen replied:
Ah thanks, I did not know that :-)

atsareg commented:
This is very smart indeed. But why not just define the file name based conditions with regular expressions ? This is way more generic and much simpler to implement. As for the Proxy plugin - I think the name of the plugin is not very appropriate, the filtering is on user identity and rights which just happens to be embedded in the Proxy. I would call it UserPlugin, for example

chaen commented:
To be franc, I did not put much effort into the file name plugin. It was just a very simple test case. Moreover, with the re module, I am afraid the logic for parsing becomes extremely complex, because the regular expression might define some reserved character. 
For the "proxy one", sure I can change the name, but User might be a bit restrictive no ? since it can also test groups, properties and VOMs.... anyone has a better name ? :-)

atsareg commented:
I will merge this PR as soon as the call to this functionality will be placed in the FileCatalog.

chaen commented:
Added the call to the FCConditionParser to the FileCatalog, corrected bugs in the FileCatalog itself, and wrote tests for its logic.
Someone please review the PR

chaen commented:
And very important, I remind that this PR requires pyparsing in the external

fstagni commented on Resources/Catalog/test/testFileCatalog.py:
```diff
@@ -0,0 +1,392 @@
+
+if __name__ == '__main__':
+  suite = unittest.defaultTestLoader.loadTestsFromTestCase( TestInitialization )
+  suite.addTest( unittest.defaultTestLoader.loadTestsFromTestCase( TestWrite ) )
```
add TestRead:
suite.addTest( unittest.defaultTestLoader.loadTestsFromTestCase( TestRead ) )

chaen replied:
Thanks, done, they pass

fstagni commented on Resources/Catalog/ConditionPlugins/FilenamePlugin.py:
```diff
@@ -0,0 +1,59 @@
+    super( FilenamePlugin, self ).__init__( conditions )
+
+
+  def eval( self, **kwargs ):
```
python eval vs custom eval... is it intended?

chaen replied:
Kind of. They semanticaly do the same. And pylint does not complain about it. If it really is a problem I can change it though.

fstagni commented on Resources/Catalog/FileCatalog.py:
```diff
@@ -209,10 +255,10 @@ def w_execute( self, *parms, **kws ):
-          failed[lfnMapDict.get( lfn, lfn )] = failed[lfn]
-        for lfn in successful:
-          successful[lfnMapDict.get( lfn, lfn )] = successful[lfn]
+        for lfn in failed.keys():
```
just double checking: do you really need .keys() here?

chaen replied:
Yes because you delete elements of the dictionnary on which you loop 

In [1]: d = dict.fromkeys(range(5))

In [2]: d
Out[2]: {0: None, 1: None, 2: None, 3: None, 4: None}

In [3]: for k in d:
   ...:     d.pop(k)
   ...:     
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-3-ce2d839d70b1> in <module>()
----> 1 for k in d:
      2     d.pop(k)
      3 

RuntimeError: dictionary changed size during iteration

fstagni commented:
Regarding pyparsing: I don't know very well the "externals" but I believe you may update
https://github.com/DIRACGrid/Externals/blob/master/ClientPackages/dirac-make.py

chaen commented:
Does that also get installed then on the server installation ?

fstagni commented:
I believe they do (since even sqlAlchemy is in there). Anyway I'm doing a PR for the server part, I'll add it there.

chaen commented:
I merged the changes of Andrei, and adapted the tests in consequence.

atsareg commented:
Review OK
